sphericart/include/dynamic_cuda.hpp

```diff
@@ -0,0 +1,357 @@
+    nvrtcAddNameExpression_t nvrtcAddNameExpression;
+    nvrtcDestroyProgram_t nvrtcDestroyProgram;
+    nvrtcGetErrorString_t nvrtcGetErrorString;
+
+    NVRTC() {
+#ifdef __linux__
+        nvrtcHandle = dlopen("libnvrtc.so", RTLD_NOW);
+#else
+#error "Platform not supported"
+#endif
+
+        if (!nvrtcHandle) {
+            throw std::runtime_error("Failed to load libnvrtc.so. Make sure it is available in "
+                                     "your $LD_LIBRARY_PATH environment variable.");
+        }
+
+        // Load NVRTC function pointers using template
+        nvrtcCreateProgram = load<nvrtcCreateProgram_t>(nvrtcHandle, "nvrtcCreateProgram");
+        nvrtcCompileProgram = load<nvrtcCompileProgram_t>(nvrtcHandle, "nvrtcCompileProgram");
+        nvrtcGetPTX = load<nvrtcGetPTX_t>(nvrtcHandle, "nvrtcGetPTX");
+        nvrtcGetPTXSize = load<nvrtcGetPTXSize_t>(nvrtcHandle, "nvrtcGetPTXSize");
+        nvrtcGetProgramLog = load<nvrtcGetProgramLog_t>(nvrtcHandle, "nvrtcGetProgramLog");
+        nvrtcGetProgramLogSize =
+            load<nvrtcGetProgramLogSize_t>(nvrtcHandle, "nvrtcGetProgramLogSize");
+        nvrtcGetLoweredName = load<nvrtcGetLoweredName_t>(nvrtcHandle, "nvrtcGetLoweredName");
+        nvrtcAddNameExpression =
+            load<nvrtcAddNameExpression_t>(nvrtcHandle, "nvrtcAddNameExpression");
+        nvrtcDestroyProgram = load<nvrtcDestroyProgram_t>(nvrtcHandle, "nvrtcDestroyProgram");
+        nvrtcGetErrorString = load<nvrtcGetErrorString_t>(nvrtcHandle, "nvrtcGetErrorString");
+    }
+
+    ~NVRTC() {
+#ifdef __linux__
+        if (nvrtcHandle) {
+            dlclose(nvrtcHandle);
+        }
+#else
+#error "Platform not supported"
+#endif
+    }
+
+    // Prevent copying
+    NVRTC(const NVRTC&) = delete;
+    NVRTC& operator=(const NVRTC&) = delete;
+
+    void* nvrtcHandle = nullptr;
+};
+
+/*
+This implements the Schawrz counter idiom to ensure propper constructor/destructor ordering - we
+don't want these classes destroyed before jax or pytorch classes/objects are.
+*/
+
+// Static memory buffer for each class
+static std::aligned_storage<sizeof(CUDART), alignof(CUDART)>::type cudartBuffer;
+static std::aligned_storage<sizeof(CUDADriver), alignof(CUDADriver)>::type cudaDriverBuffer;
+static std::aligned_storage<sizeof(NVRTC), alignof(NVRTC)>::type nvrtcBuffer;
+
+// global references, use inline instead of extern as we want to define the implementation here
+inline CUDART& cudart = reinterpret_cast<CUDART&>(cudartBuffer);
+inline CUDADriver& cudadriver = reinterpret_cast<CUDADriver&>(cudaDriverBuffer);
+inline NVRTC& nvrtc = reinterpret_cast<NVRTC&>(nvrtcBuffer);
+
+static int nifty_counter = 0;
+
+// Counter class for initializing and destroying static objects
+static struct CUDAInitializer {
+
+    CUDAInitializer() {
+        if (nifty_counter++ == 0) {
+            new (&cudart) CUDART();
+            new (&cudadriver) CUDADriver();
+            new (&nvrtc) NVRTC();
```

Comment:
I'm not sure I understand how this will work on systems without a CUDA SDK: the os will try to create these when loading `sphericart.so`, fail to load `libcuda.so` and crash with the errors above

Reply:
you are right!

maybe we should "do nothing" if the libraries aren't found - no warnings etc, and then rely on the sphericart_cuda.cpp classes to throw the runtime errors, that will allow us to do both CPU and GPU then.

Reply:
I've also added a brief comment on the remove_pointer structs